core/serialize/serialize.cc

```diff
@@ -1035,19 +981,90 @@ void SerializerImpl::unpickleGS(UnPickler &p, GlobalState &result) {
         Timer timeit(result.tracer(), "moving");
         result.fileRefByPath = std::move(fileRefByPath);
         result.files = std::move(files);
-        result.utf8Names = std::move(utf8Names);
-        result.constantNames = std::move(constantNames);
-        result.uniqueNames = std::move(uniqueNames);
         result.classAndModules = std::move(classAndModules);
         result.methods = std::move(methods);
         result.fields = std::move(fields);
         result.typeArguments = std::move(typeArguments);
         result.typeMembers = std::move(typeMembers);
-        result.namesByHash = std::move(namesByHash);
     }
     result.sanityCheck();
 }
 
+void SerializerImpl::pickleNameTable(Pickler &p, const GlobalState &gs) {
+    p.putU4(gs.utf8Names.size());
+    for (const auto &n : gs.utf8Names) {
+        pickle(p, n);
+    }
+    p.putU4(gs.constantNames.size());
+    for (const auto &n : gs.constantNames) {
+        pickle(p, n);
+    }
+    p.putU4(gs.uniqueNames.size());
+    for (const auto &n : gs.uniqueNames) {
+        pickle(p, n);
+    }
+
+    p.putU4(gs.namesByHash.size());
+    for (const auto &s : gs.namesByHash) {
+        p.putU4(s.hash);
+        p.putU4(s.rawId);
+    }
+}
+
+void SerializerImpl::unpickleNameTable(UnPickler &p, GlobalState &result) {
+    vector<UTF8Name> utf8Names(std::move(result.utf8Names));
+    utf8Names.clear();
+    vector<ConstantName> constantNames(std::move(result.constantNames));
+    constantNames.clear();
+    vector<UniqueName> uniqueNames(std::move(result.uniqueNames));
+    uniqueNames.clear();
+    vector<GlobalState::Bucket> namesByHash(std::move(result.namesByHash));
+    namesByHash.clear();
```

Comment:
This is some weird code.

Is there a reason why we don't take `auto &` to these four fields in `result`, call `.clear()` on them, and skip the final moves back into the fields?

Reply:
Nope, I was preserving the existing implementation when moving it to a new function to ensure that it was clear where it came from. I'm happy to refactor it.